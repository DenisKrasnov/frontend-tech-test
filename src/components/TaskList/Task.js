import React from "react";
import { shape, string, func } from "prop-types";

import Icon from "../Icon";
import styles from "./Task.scss";

const Task = props => {
  const onEdit = () => {
    const { task, openEditModal } = props;
    openEditModal(task);
  };

  const onDelete = () => {
    const { task, deleteTask, clearError } = props;
    clearError();
    deleteTask({ id: task.id });
  };

  const { task } = props;

  return (
    <div className={styles.taskContainer}>
      <section className={styles.task}>
        <h4 className={styles.title}>{task.title}</h4>
        <p className={styles.description}>{task.description}</p>
      </section>
      <div className={styles.actions}>
        <button className={styles.iconButton} onClick={onDelete} type="button">
          <Icon glyph="trash" />
        </button>
        <button className={styles.iconButton} onClick={onEdit} type="button">
          <Icon glyph="pencil" />
        </button>
      </div>
    </div>
  );
};

Task.propTypes = {
  task: shape({
    id: string.isRequired,
    title: string,
    description: string,
  }).isRequired,
  openEditModal: func.isRequired,
  deleteTask: func.isRequired,
  clearError: func.isRequired,
};

export default Task;
