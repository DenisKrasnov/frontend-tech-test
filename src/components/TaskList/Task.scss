.taskContainer {
  background: #fff;
  box-shadow: 0 1px 5px $shadow-color;
  margin: 10px;
}

.task {
  display: flex;
  flex-direction: column;
  padding: 0 20px;
  width: 300px;
  height: 210px;
}

.title {
  margin-block-end: 10px;
  overflow-wrap: break-word;
}

.description {
  margin: 0;
  overflow: hidden;
  overflow-wrap: break-word;
}

.iconButton {
  display: inline-block;
  padding: 4px;
  border: 0;
  outline: 0;
  margin-left: 10px;

  &:hover {
    svg {
      fill: $icon-hover-color;
    }
  }
}

.actions {
  display: flex;
  flex-direction: row-reverse;
  padding: 10px;
  background-color: #fff;
  border-top: 1px dashed $border-color;
}
