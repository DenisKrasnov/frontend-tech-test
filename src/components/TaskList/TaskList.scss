.TaskList {
  position: relative;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  width: 100%;
  padding: 0 10px 20px 10px;
}

.counterWrapper {
  position: absolute;
  top: -20px;
  font-weight: 600;
}
