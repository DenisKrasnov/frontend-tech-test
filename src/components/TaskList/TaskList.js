import React, { Component } from "react";
import { connect } from "react-redux";
import { array, func } from "prop-types";

import {
  fetchTasks,
  getTasks,
  updateTask,
  deleteTask,
} from "../../services/redux/tasks";
import { maxLengths } from "../../utilities/utilities";

import Counter from "../Counter";
import TaskModal from "../TaskModal";
import Task from "./Task";

import styles from "./TaskList.scss";

export class TaskList extends Component {
  static propTypes = {
    updateTask: func.isRequired,
    fetchTasks: func.isRequired,
    tasks: array.isRequired,
    clearError: func.isRequired,
    deleteTask: func.isRequired,
  };

  state = {
    showModal: false,
    task: null,
  };

  componentDidMount() {
    this.props.fetchTasks();
  }

  openEditModal = task => this.setState({ showModal: true, task: task });

  closeModal = () => this.setState({ showModal: false, task: null });

  editTask = values => {
    const { task } = this.state;
    if (
      values.title === task.title &&
      values.description === task.description
    ) {
      this.closeModal();
      return;
    }
    this.props.updateTask({ ...values, id: task.id });
    this.closeModal();
  };

  render() {
    const { showModal, task } = this.state;
    const { tasks, deleteTask, clearError } = this.props;

    return (
      <div className={styles.TaskList}>
        <span className={styles.counterWrapper}>
          <Counter valueLength={tasks.length} maxLength={maxLengths.tasks} />
        </span>
        {tasks.map(task => (
          <Task
            key={task.id}
            task={task}
            openEditModal={this.openEditModal}
            deleteTask={deleteTask}
            clearError={clearError}
          />
        ))}
        <TaskModal
          isOpen={showModal}
          onClose={this.closeModal}
          editTask={this.editTask}
          task={task}
        />
      </div>
    );
  }
}

const stateToProps = state => getTasks(state);
const dispatchToProps = { fetchTasks, updateTask, deleteTask };

export default connect(
  stateToProps,
  dispatchToProps
)(TaskList);
