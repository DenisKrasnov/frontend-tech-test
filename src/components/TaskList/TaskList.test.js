import React from "react";
import { shallow } from "enzyme";

import { TaskList } from "./TaskList";

const noop = () => {};
const tasks = [
  {
    id: "2",
    title: "__TITLE__",
    description: "__DESCRIPTION__",
  },
];

describe("TaskList", () => {
  it("renders", () => {
    const component = shallow(
      <TaskList
        updateTask={noop}
        fetchTasks={noop}
        clearError={noop}
        deleteTask={noop}
        tasks={tasks}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("triggers the 'updateTask' and 'fetchTasks'", () => {
    const updateTaskSpy = jest.fn();
    const fetchTasksSpy = jest.fn();
    const values = {
      title: "__SOME_TITLE__",
      description: "__SOME_DESCRIPTION__",
    };
    const component = shallow(
      <TaskList
        updateTask={updateTaskSpy}
        fetchTasks={fetchTasksSpy}
        clearError={noop}
        deleteTask={noop}
        tasks={tasks}
      />
    );

    // check 'fetchTasks'
    expect(fetchTasksSpy).toHaveBeenCalledTimes(1);

    // check 'updateTask'
    component.instance().openEditModal(tasks[0]);
    const taskId = component.state().task.id;
    component.instance().editTask(values);
    expect(updateTaskSpy).toHaveBeenCalledWith({ ...values, id: taskId });
  });

  it("tests the state when the 'openEditModal' and 'closeModal' are triggered", () => {
    const component = shallow(
      <TaskList
        updateTask={noop}
        fetchTasks={noop}
        clearError={noop}
        deleteTask={noop}
        tasks={tasks}
      />
    );

    component.instance().openEditModal(tasks[0]);

    expect(component.state().task).toEqual(tasks[0]);
    expect(component.state().showModal).toBe(true);

    component.instance().closeModal(tasks[0]);

    expect(component.state().task).toBeNull();
    expect(component.state().showModal).toBe(false);
  });
});
