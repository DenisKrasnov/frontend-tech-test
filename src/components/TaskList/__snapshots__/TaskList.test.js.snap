// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList renders 1`] = `
<div
  className="TaskList"
>
  <span
    className="counterWrapper"
  >
    <Counter
      maxLength={50}
      valueLength={1}
    />
  </span>
  <Task
    clearError={[Function]}
    deleteTask={[Function]}
    key="2"
    openEditModal={[Function]}
    task={
      Object {
        "description": "__DESCRIPTION__",
        "id": "2",
        "title": "__TITLE__",
      }
    }
  />
  <TaskModal
    editTask={[Function]}
    isOpen={false}
    onClose={[Function]}
    task={null}
  />
</div>
`;
